File: CodalityTestHarness/CodifyTestHarness/5. Prefix Sums/GeonomicRangeQuery.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace CodalityTestLibrary
{
    /// <summary>
    /// Write a function that, given a non-empty zero-indexed string S 
    /// consisting of N characters and two non-empty zero-indexed arrays 
    /// P and Q consisting of M integers, returns an array consisting 
    /// of M integers specifying the consecutive answers to all queries
    /// </summary>
    public class GeonomicRangeQuery
    {
        public int[] solution(string sequence, int[] leftArray, int[] rightArray)
        {
            // Gaurd against empty arrays
            if(leftArray.Length == 0 ){
                throw new ArgumentNullException(nameof(leftArray)); 
            }

            if (rightArray.Length == 0)
            {
                throw new ArgumentNullException(nameof(rightArray));
            }

            // Store the subsequence min values
            List<int> returnList = new List<int>();

            // Use the first array as our enumerator
            for (int i = 0; i < leftArray.Length; i++)
            {
                // Get the substring
                string subsequence = sequence.Substring(leftArray[i], (rightArray[i] - leftArray[i]) + 1);

                // Get the associated enum, find the lowest value
                returnList.Add(subsequence.Select(x => (int)Enum.Parse(typeof(Nucleotides), x.ToString())).Min());
            }

            return returnList.ToArray();
        }

        /// <summary>
        /// Nucleotides values
        /// </summary>
        enum Nucleotides
        {
            A = 1,
            C = 2,
            G = 3,
            T = 4
        }
    }
}

File: Python/5. Prefix Sums/GeonomicRangeQuery.py
# Write a function that, given a non-empty zero-indexed string S 
# consisting of N characters and two non-empty zero-indexed arrays 
# P and Q consisting of M integers, returns an array consisting 
# of M integers specifying the consecutive answers to all queries

class GeonomicRangeQuery(object):

    def solution(self, sequence, leftArray, rightArray):
        
        #veto empty arrays
        if len(leftArray) == 0 or len(rightArray) == 0:
            return 0

        # store return values and setup a lookup
        returnList = []
        lookup = {"A": 1, "C": 2, "G": 3, "T": 4}

        # iterate on one of the lists
        for index, value in enumerate(leftArray):
            # get the subsequence based on the two array points
            subsequence = sequence[value:(rightArray[index] + 1)]
            
            # do some groovy python list querying to get the values from 
            # the lookup and also find the smallest value
            minvalue = min([lookup[x] for x in subsequence])
            returnList.append(minvalue)
        
        # return
        return returnList
